File: tests/cpp/test_runtime_spec_conformance.cpp
#include <hgraph/lib/std/std_operators.h>
#include <hgraph/lib/testing/check_output.h>
#include <hgraph/lib/testing/eval_node.h>
#include <hgraph/types/graph_wiring.h>
#include <hgraph/types/static_node.h>
#include <hgraph/types/time_series/ts_delta.h>
#include <catch2/catch_test_macros.hpp>

namespace
{
    using namespace hgraph;
    using Dict = TSD<Str, TS<Int>>;

    struct MembershipSource
    {
        static void eval(In<"step", TS<Int>> step, Out<Dict> out)
        {
            const Str key{"X"};
            switch (step.value())
            {
                case 1: static_cast<void>(out.at(key)); break;
                case 2: case 6: out.set(key, Int{7}); break;
                case 3: {
                    auto child = out.at(key);
                    static_cast<void>(child.begin_mutation(child.evaluation_time()).invalidate());
                    break;
                }
                case 4: static_cast<void>(out.erase(key)); break;
            }
        }
    };

    struct ObserveMembership
    {
        static void eval(In<"step", TS<Int>> step,
                         In<"ts", Dict, InputValidity::Unchecked, InputActivity::Passive> ts,
                         Out<TS<Int>> out)
        {
            const auto n = step.value();
            std::size_t added = 0, removed = 0;
            for ([[maybe_unused]] const auto &key : ts.added_keys()) { ++added; }
            for ([[maybe_unused]] const auto &key : ts.removed_keys()) { ++removed; }
            CHECK(added == (n == 1 || n == 6 ? 1 : 0));
            CHECK(removed == (n == 4 ? 1 : 0));
            CHECK(ts.data_view().key_set().modified(ts.evaluation_time()) == (n == 1 || n == 4 || n == 6));
            auto key_set = ts.data_view().key_set();
            if (key_set.modified(ts.evaluation_time()))
            {
                CHECK(std::ranges::distance(key_set.added()) == (n == 4 ? 0 : 1));
                CHECK(std::ranges::distance(key_set.removed()) == (n == 4 ? 1 : 0));
                const auto delta = key_set.base().delta_value(ts.evaluation_time());
                CHECK(delta.as_indexed().at(0).as_set().size() == (n == 4 ? 0 : 1));
                CHECK(delta.as_indexed().at(1).as_set().size() == (n == 4 ? 1 : 0));
            }
            out.set(n);
        }
    };

    struct MembershipGraph
    {
        static Port<TS<Int>> compose(Wiring &w, Port<TS<Int>> step)
        {
            return wire<ObserveMembership>(w, step, wire<MembershipSource>(w, step)).as<TS<Int>>();
        }
    };

    template <bool Right>
    struct DictionarySource
    {
        static void eval(In<"step", TS<Int>> step, Out<Dict> out)
        {
            if (step.value() == 0)
            {
                out.set(Str{Right ? "Y" : "X"}, Int{Right ? 2 : 1});
                out.set(Str{"Z"}, Int{Right ? 3 : 9});
            }
            if constexpr (Right) { if (step.value() == 1) { out.set(Str{"Y"}, Int{4}); } }
        }
    };

    struct SelectDictionary
    {
        static void eval(In<"step", TS<Int>> step,
                         In<"a", REF<Dict>, InputActivity::Passive> a,
                         In<"b", REF<Dict>, InputActivity::Passive> b, Out<REF<Dict>> out)
        {
            if (step.value() == 0 || step.value() == 4) { out.set(a.value()); }
            if (step.value() == 2) { out.set(b.value()); }
            if (step.value() == 3) { out.set(TimeSeriesReference::empty()); }
        }
    };

    struct ObserveDictionary
    {
        static void eval(In<"step", TS<Int>> step,
                         In<"ts", Dict, InputValidity::Unchecked, InputActivity::Passive> ts,
                         Out<TS<Int>> out)
        {
            const auto n = step.value();
            if (n == 2 || n == 4)
            {
                for (const auto &[key, child] : ts.items())
                {
                    static_cast<void>(key);
                    CHECK(child.modified());
                    CHECK(child.last_modified_time() == ts.evaluation_time());
                }
            }
            if (n == 2)
            {
                CHECK(capture_delta(ts.base()).equals(dict_delta<Str, TS<Int>>({{"Y", 4}, {"Z", 3}}, {"X"})));
                std::size_t removed = 0;
                for (const auto &[key, child] : ts.removed_items())
                {
                    CHECK(key.checked_as<Str>() == "X");
                    CHECK(child.value() == 1);
                    ++removed;
                }
                CHECK(removed == 1);
            }
            if (n == 3)
            {
                CHECK_FALSE(ts.valid());
                CHECK(ts.modified());
                CHECK(ts.last_modified_time() == MIN_DT);
                CHECK(capture_delta(ts.base()).equals(dict_delta<Str, TS<Int>>({}, {"Y", "Z"})));
                std::size_t removed = 0;
                for (const auto &[key, child] : ts.removed_items())
                {
                    CHECK(child.value() == (key.checked_as<Str>() == "Y" ? 4 : 3));
                    ++removed;
                }
                CHECK(removed == 2);
            }
            out.set(n);
        }
    };

    struct DictionaryGraph
    {
        static Port<TS<Int>> compose(Wiring &w, Port<TS<Int>> step)
        {
            auto a = wire<DictionarySource<false>>(w, step);
            auto b = wire<DictionarySource<true>>(w, step);
            return wire<ObserveDictionary>(w, step, wire<SelectDictionary>(w, step, a, b)).as<TS<Int>>();
        }
    };

    struct SelectScalar
    {
        static void eval(In<"step", TS<Int>> step, In<"value", REF<TS<Int>>> value, Out<REF<TS<Int>>> out)
        {
            out.set(step.value() == 0 ? value.value() : TimeSeriesReference::empty());
        }
    };

    struct ObserveScalar
    {
        static void eval(In<"step", TS<Int>> step,
                         In<"value", TS<Int>, InputValidity::Unchecked, InputActivity::Passive> value, Out<TS<Int>> out)
        {
            CHECK(value.valid() == (step.value() == 0));
            CHECK(value.modified() == (step.value() == 0));
            CHECK(value.last_modified_time() == (step.value() == 0 ? MIN_ST : MIN_DT));
            out.set(step.value());
        }
    };

    struct ScalarReferenceGraph
    {
        static Port<TS<Int>> compose(Wiring &w, Port<TS<Int>> step, Port<TS<Int>> value)
        {
            return wire<ObserveScalar>(w, step, wire<SelectScalar>(w, step, value)).as<TS<Int>>();
        }
    };

    struct SaveReference
    {
        static void eval(In<"step", TS<Int>> step,
                         In<"ref", REF<TS<Int>>, InputValidity::Unchecked, InputActivity::Passive> ref,
                         State<TimeSeriesReference> saved, Out<REF<TS<Int>>> out)
        {
            if (step.value() == 0) { saved.set(ref.value()); }
            if (step.value() == 0 || step.value() == 3 || step.value() == 5) { out.set(saved.get()); }
        }
    };

    struct ObserveSavedReference
    {
        static void eval(In<"step", TS<Int>> step,
                         In<"ts", TS<Int>, InputValidity::Unchecked, InputActivity::Passive> ts,
                         Out<TS<Int>> out)
        {
            CHECK(ts.modified() == (step.value() == 0));
            out.set(ts.valid() ? ts.value() : Int{-1});
        }
    };

    struct SavedReferenceGraph
    {
        static Port<TS<Int>> compose(Wiring &w, Port<TS<Int>> step, Port<Dict> values)
        {
            auto ref = wire<stdlib::getitem_>(w, values, Str{"X"});
            return wire<ObserveSavedReference>(w, step, wire<SaveReference>(w, step, ref)).as<TS<Int>>();
        }
    };
}

TEST_CASE("Runtime rules: membership and key set observations through native wiring", "[runtime-contract]")
{
    CHECK_OUTPUT(hgraph::testing::eval_node<MembershipGraph>({0, 1, 2, 3, 4, 5, 6, 7}), {0, 1, 2, 3, 4, 5, 6, 7});
}

TEST_CASE("Runtime rules: dictionary sampling and withdrawal through native wiring", "[runtime-contract]")
{
    CHECK_OUTPUT(hgraph::testing::eval_node<DictionaryGraph>({0, 1, 2, 3, 4, 5}), {0, 1, 2, 3, 4, 5});
}

TEST_CASE("Runtime rules: saved reference expires while dictionary reclamation remains lazy", "[runtime-contract]")
{
    using namespace hgraph;
    using namespace hgraph::testing;
    stdlib::register_standard_operators();
    CHECK_OUTPUT(eval_node<SavedReferenceGraph>(values<Int>(0, 1, 2, 3, 4, 5, 6),
        values<Value>(dict_delta<Str, TS<Int>>({{"X", 7}}), none,
            dict_delta<Str, TS<Int>>({}, {"X"}), none, dict_delta<Str, TS<Int>>({{"X", 9}}), none, none)),
        {7, 7, 7, -1, -1, -1, -1});
}

TEST_CASE("Runtime rules: scalar reference withdrawal resets observed time through native wiring", "[runtime-contract]")
{
    using namespace hgraph;
    using namespace hgraph::testing;
    CHECK_OUTPUT(eval_node<ScalarReferenceGraph>(values<Int>(0, 1), values<Int>(7, none)), {0, 1});
}

File: python/tests/test_runtime_spec_conformance.py
"""Accepted runtime rules from runtime_spec/validation.md, DV-01/02/05/06/08/10."""
import hgraph as hg
from hgraph.test import eval_node


def test_dictionary_membership_and_key_set_clock():
    observations = []

    @hg.compute_node
    def source(step: hg.TS[int], _output: hg.TSD = None) -> hg.TSD[str, hg.TS[int]]:
        if step.value == 1:
            _output.get_or_create("X")
        elif step.value in (2, 6):
            _output.get_or_create("X").value = 7
        elif step.value == 3:
            _output["X"].invalidate()
        elif step.value == 4:
            del _output["X"]

    @hg.compute_node(valid=("step",), active=("step",))
    def observe(step: hg.TS[int], ts: hg.TSD[str, hg.TS[int]]) -> hg.TS[int]:
        observations.append((set(ts.added_keys()), set(ts.removed_keys()), ts.key_set.modified))
        if ts.key_set.modified:
            expected = {hg.Removed("X")} if step.value == 4 else {"X"}
            assert set(ts.key_set.delta_value) == expected
        return step.value

    @hg.graph
    def run(step: hg.TS[int]) -> hg.TS[int]:
        return observe(step, source(step))

    eval_node(run, list(range(8)))
    assert observations == [
        (set(), set(), False), ({"X"}, set(), True),
        (set(), set(), False), (set(), set(), False),
        (set(), {"X"}, True), (set(), set(), False),
        ({"X"}, set(), True), (set(), set(), False),
    ]


def test_dictionary_reference_sampling_and_withdrawal():
    observations = []

    @hg.compute_node(valid=("selector",), active=("selector",))
    def choose(selector: hg.TS[int], a: hg.REF[hg.TSD[str, hg.TS[int]]],
               b: hg.REF[hg.TSD[str, hg.TS[int]]]) -> hg.REF[hg.TSD[str, hg.TS[int]]]:
        return [a.value, b.value, hg.TimeSeriesReference.make()][selector.value]

    @hg.compute_node(valid=("step",), active=("step",))
    def observe(step: hg.TS[int], ts: hg.TSD[str, hg.TS[int]]) -> hg.TS[int]:
        observations.append({
            "valid": ts.valid,
            "modified": ts.modified,
            "delta": dict(ts.delta_value) if ts.modified else None,
            "removed": {k: v.value for k, v in ts.removed_items()},
            "times": {k: v.last_modified_time for k, v in ts.items()},
        })
        return step.value

    @hg.graph
    def run(step: hg.TS[int], a: hg.TSD[str, hg.TS[int]], b: hg.TSD[str, hg.TS[int]],
            selector: hg.TS[int]) -> hg.TS[int]:
        return observe(step, choose(selector, a, b))

    eval_node(run, list(range(6)), [{"X": 1, "Z": 9}, None, None, None, None, None],
              [{"Y": 2, "Z": 3}, {"Y": 4}, None, None, None, None], [0, None, 1, 2, 0, None])
    assert [o["delta"] for o in observations] == [
        {"X": 1, "Z": 9}, None, {"X": hg.REMOVE, "Y": 4, "Z": 3},
        {"Y": hg.REMOVE, "Z": hg.REMOVE}, {"X": 1, "Z": 9}, None,
    ]
    assert observations[2]["removed"] == {"X": 1}
    assert observations[3]["removed"] == {"Y": 4, "Z": 3}
    assert observations[3]["valid"] is False
    assert observations[3]["modified"] is True
    assert observations[2]["times"] == {"Y": hg.MIN_ST + 2 * hg.MIN_TD, "Z": hg.MIN_ST + 2 * hg.MIN_TD}
    assert observations[4]["times"] == {"X": hg.MIN_ST + 4 * hg.MIN_TD, "Z": hg.MIN_ST + 4 * hg.MIN_TD}


def test_scalar_reference_unbind_resets_observed_time():
    observations = []

    @hg.compute_node
    def choose(step: hg.TS[int], a: hg.REF[hg.TS[int]]) -> hg.REF[hg.TS[int]]:
        return a.value if step.value == 0 else hg.TimeSeriesReference.make()

    @hg.compute_node(valid=("step",), active=("step",))
    def observe(step: hg.TS[int], ts: hg.TS[int]) -> hg.TS[int]:
        observations.append((ts.valid, ts.modified, ts.last_modified_time))
        return step.value

    @hg.graph
    def run(step: hg.TS[int], a: hg.TS[int]) -> hg.TS[int]:
        return observe(step, choose(step, a))

    eval_node(run, [0, 1], [7, None])
    assert observations == [(True, True, hg.MIN_ST), (False, False, hg.MIN_DT)]


def test_saved_reference_expires_before_next_dictionary_mutation():
    observations = []

    @hg.compute_node(valid=("step",), active=("step",))
    def save(step: hg.TS[int], ref: hg.REF[hg.TS[int]], _state: hg.STATE = None) -> hg.REF[hg.TS[int]]:
        if step.value == 0:
            _state.saved = ref.value
        if step.value in (0, 3, 5):
            return _state.saved

    @hg.compute_node(valid=("step",), active=("step",))
    def observe(step: hg.TS[int], ts: hg.TS[int]) -> hg.TS[int]:
        observations.append((ts.value if ts.valid else None, ts.modified))
        return step.value

    @hg.graph
    def run(step: hg.TS[int], values: hg.TSD[str, hg.TS[int]]) -> hg.TS[int]:
        return observe(step, save(step, values["X"]))

    eval_node(run, list(range(7)), [{"X": 7}, None, {"X": hg.REMOVE}, None, {"X": 9}, None, None])
    assert observations == [(7, True), (7, False), (7, False), (None, False),
                            (None, False), (None, False), (None, False)]
